### PR TITLE
refactor: func.d add canInferAttributes() and initInferAttributes()

### PR DESCRIPTION
**Pull request by WalterBright**
Done to make the logic easier to follow.

### CODE REVIEW

**Comment by UplinkCoder** [comment]
Looks good!

**Comment by ZombineDev** [comment]
LGTM too

File: src/func.d [hunk 1]
```diff
@@ -3029,6 +3008,46 @@ extern (C++) class FuncDeclaration : Declaration
...
+     * Decide if attributes for this function can be inferred from examining
+     * the function body.
+     * Returns:
+     *  true if can
+     */
+    private final bool canInferAttributes(Scope* sc)
+    {
+        TemplateInstance ti;
+        return (fbody && !isVirtualMethod() &&
+                        /********** this is for backwards compatibility for the moment ********/
```
Comment (andralex) [inline src/func.d]: "this" meaning "the two checks above" or "the three checks below"? Or something else?

File: src/func.d [hunk 2]
```diff
@@ -3029,6 +3008,46 @@ extern (C++) class FuncDeclaration : Declaration
...
+            (sc.func && (!isMember() || sc.func.isSafeBypassingInference() && !isInstantiated()) ||
```
Comment (andralex) [inline src/func.d]: superfluous parens but passons

File: src/func.d (hunk 1 above)
Comment (WalterBright) [inline src/func.d]: Below, and the beginning/end of the comments line up with what it refers to. In the next PR, I plan to straighten out the logic here. But in order to make this refactor easy to review, I stayed with it being a mere cut/paste of the original code.

File: src/func.d (hunk 2 above)
Comment (WalterBright) [inline src/func.d]: same reply as above

File: src/func.d (hunk 1 above)
Comment (andralex) [inline src/func.d]: Well I'd rather have had that in the comment proper. But passons.

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**